renku/core/models/migration.py

```diff
@@ -0,0 +1,218 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2017-2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Model used for migrations."""
+
+import os
+
+from renku.core.models.git import get_user_info
+from renku.core.models.jsonld import write_yaml
+from renku.core.models.projects import generate_project_id
+from renku.core.utils.migrate import read_jsonld_yaml
+
+
+def _to_dict(value):
+    if isinstance(value, Base):
+        return value.metadata
+    if isinstance(value, dict):
+        return value
+    raise ValueError
+
+
+def _DatasetFile_to_dict(value):
+    from renku.core.models.datasets import DatasetFile
+
+    if isinstance(value, DatasetFile):
+        return value.as_jsonld()
+    if isinstance(value, dict):
+        return value
+    raise ValueError
+
+
+class Base:
```

Comment:
We still have the issue here that we use one schema (the ones in this file) for all migrations, but schemas might change potentially leading to issues further down the line. I think it would be safest to have the schemas inside the individual migrations, even if it is code duplication. We could add the Schema to the first migration that uses it, import if from there in successive migrations if it didn't change, but add a modified copy to a migration if the schema changed.

And we could use calamus for this if we added a flag to calamus to not create an instance but instead just output a dict (like base marshmallow does). Maybe an option like `create_instance=False` passed to the schema?

And for instance workflows are represented as a flattened list instead of a nested jsonld structure, so this code would not be able to handle that, but calamus can.

Reply:
I've ended up adding a generic base class that gets its attributes from the schema that is used for loading.